src/Microsoft.ML.Parquet/ParquetLoader.cs

```diff
@@ -390,11 +391,32 @@ public Cursor(ParquetLoader parent, Func<int, bool> predicate, IRandom rand)
                     Columns = _loader._columnsLoaded.Select(i => i.Name).ToArray()
                 };
 
-                int numBlocks = (int)Math.Ceiling(((decimal)parent.GetRowCount() / _readerOptions.Count));
-                int[] blockOrder = _rand == null ? Utils.GetIdentityPermutation(numBlocks) : Utils.GetRandomPermutation(rand, numBlocks);
-                _blockEnumerator = blockOrder.GetEnumerator();
+                // The number of blocks is calculated based on the specified rows in a block (defaults to 1M).
+                // Since we want to shuffle the blocks in addition to shuffling the rows in each block, checks
+                // are put in place to ensure we can produce a shuffle order for the blocks.
+                int numBlocks;
+                int[] blockOrder;
+                try
+                {
+                    numBlocks = checked((int)Math.Ceiling(((decimal)parent.GetRowCount() / _readerOptions.Count)));
```

Comment:
>checked [](start = 32, length = 7)

consider using numblocks  as long and then just use an if statement if it is above int.max
Why need to throw and then catch?

Reply:
Yes, exception driven computation is kind of one of those mechanisms best left as a last resort.

---
In reply to: [187492091](https://github.com/dotnet/machinelearning/pull/120#discussion_r187492091) [](ancestors = 187492091)

Reply:
Fixed!